### ISSUE TITLE
az aks get-credentials should return unique credentials for each user instead of returning shared credentials

### ISSUE THREAD

**Issue by DaveAurionix** [Azure/AKS#397]
For General Availability of AKS, will `az aks get-credentials` enforce kubectl to connect with credentials unique to each AAD user logged in via `az login` instead of returning shared credentials that are the same for all users as it does at the moment?<br><br>

I'm concerned that production clusters need to be secure at General Availability such that kubectl credentials to access the cluster can't walk out of the organisation. Ideally, if an AAD account is disabled, the key in the *.kube\config* file should no longer allow cluster access. This might be RBAC-AAD integration, or it might be something simpler for GA.  One compromise might be to continue to use shared credentials (for now) but create the ability to easily cycle/refresh the credentials that `az aks get-credentials` returns so that older *.kube\config* file contents can no longer be used to access the cluster.<br><br>

Alternatively, is there something I've missed with AKS that can stop *anyone* being able to access the cluster *from anywhere* as long as they have a copy of the *currently non-revocable* credentials in that file?

(As mentioned @neilpeterson)

**Comment by JackQuincy** [comment]
@amanohar @sauryadas

**Comment by amanohar** [comment]
@DaveAurionix 

A successful AKS cluster create configures two kubconfigs with different roles: 
1. Cluster Admin kubeconfig
2. Cluster User kubeconfig 

Customers can get these kubeconfigs via **POST** action: 

`https://management.azure.com/subscriptions/${SUB_ID}/resourcegroups/${RG_NAME}/providers/Microsoft.ContainerService/managedClusters/${AKS_NAME}/accessProfiles/${ROLE_NAME}/listcredential?api-version=2018-03-31 `

 
Substitute ${ROLE_NAME} with: `clusterAdmin` or `clusterUser` role names.  

 
The **POST** action enables subscription admin to control access to sensitive information like kubeconfigs. Users in ARM may be associated with a set of roles which control access to resources or paths within a resource. For example, the `Owner` role has access to `*` and the `Reader` role has access to `*/read`. This allows quick and easy access management for customers (and in this case RBAC kubeconfigs) 
 

> Reference Material: 
> https://docs.microsoft.com/en-us/azure/active-directory/role-based-access-built-in-roles 
> https://docs.microsoft.com/en-us/azure/active-directory/role-based-access-control-custom-roles 

`clusterAdmin` kubconfig: This is the admin kubeconfig that has full access to the cluster (can be RBACed as mentioned above). Using this kubeconfig cluster admin can create roles and rolebindings to associate AAD groups for Kubernetes RBACing. 

`clusterUser` kubconfig: Cluster user kubeconfig is setup to authenticate and authorize via AAD. By default, an AAD user has no access to the cluster and authorization to an AKS cluster will only succeed if AKS cluster admin has setup the right role bindings for the user. 

In summary yes, az aks get-credentials will return a kubeconfig (for clusterUser role) and `kubectl` will connect with credentials unique to each AAD user logged in via device login the first time a `kubectl` command is run or when AAD token expires. So `clusterUser` kubeconfig will not return shared credentials that are the same for all users.

Also, as mentioned above `clusterAdmin` will have full access to the cluster like it does today (this can also be RBACed as described above). This is required to bootstrap the cluster for other AAD users for `clusterUser` roles.

**Comment by holmesb** [comment]
"As long as the associated AAD group is updated to remove the user or if user leaves that organization they will not be able to access the cluster." - Immediate access revocation, or next time AAD token expires?

**Comment by DaveAurionix** [comment]
@amanohar Thank you for such a detailed answer.  This sounds really good.  AKS is working out really well for us.

@holmesb It's on my list to experiment with the AAD integration more, will post back when I can confirm findings (& understanding).

**Comment by badalk** [comment]
@amanohar when you say "can be RBACed as mentioned above" - which section are you referring to? We would also like to disable az aks get-credentials with --admin option. I did not understand clearly though on how do you achieve that?

**Comment by holmesb** [comment]
+1 for somehow disabling --admin.  In many orgs, Contributor (or greater) rights are set at subscription level.  Removing is impractical or a long-term goal.  We need an interim solution.

**Comment by jimwheeler** [comment]
It bothers me that Microsoft has not answered this.  That indicates to me that this is in fact a huge security hole.  Come on Microsoft.... how do we fix this?

"+1 for somehow disabling --admin. In many orgs, Contributor (or greater) rights are set at subscription level. Removing is impractical or a long-term goal. We need an interim solution. --admin is a massive back-door which breaks our security efforts."

**Comment by danisabad** [comment]
> "As long as the associated AAD group is updated to remove the user or if user leaves that organization they will not be able to access the cluster." - Immediate access revocation, or next time AAD token expires?

Not quite immediate, rather takes ~5 minutes for me.

Question about disabling admin account **still valid**.

**Comment by zuehlke-dea** [comment]
Hi, is there any way to reset the clusterAdmin credentials? Currently the sent certificates are valid for 2 years and we have people left the company with theoretically still access with this certificate. When I get this right only with clusterUser a AAD check is done, so when having the clsuterAdmin credentials you can access the clusters even you AAD account is locked.

**Comment by asubmani** [comment]
> @DaveAurionix
> 
> A successful AKS cluster create configures two kubconfigs with different roles:
> 
> 1. Cluster Admin kubeconfig
> 2. Cluster User kubeconfig
> 
> Customers can get these kubeconfigs via **POST** action:
> 
> `https://management.azure.com/subscriptions/${SUB_ID}/resourcegroups/${RG_NAME}/providers/Microsoft.ContainerService/managedClusters/${AKS_NAME}/accessProfiles/${ROLE_NAME}/listcredential?api-version=2018-03-31 `
> 
> Substitute ${ROLE_NAME} with: `clusterAdmin` or `clusterUser` role names.
> 
> The **POST** action enables subscription admin to control access to sensitive information like kubeconfigs. Users in ARM may be associated with a set of roles which control access to resources or paths within a resource. For example, the `Owner` role has access to `*` and the `Reader` role has access to `*/read`. This allows quick and easy access management for customers (and in this case RBAC kubeconfigs)
> 
> > Reference Material:
> > https://docs.microsoft.com/en-us/azure/active-directory/role-based-access-built-in-roles
> > https://docs.microsoft.com/en-us/azure/active-directory/role-based-access-control-custom-roles
> 
> `clusterAdmin` kubeconfig: This is the admin kubeconfig that has full access to the cluster (can be RBACed as mentioned above). Using this kubeconfig cluster admin can create roles and rolebindings (to associate AAD groups with Kubernetes roles) for Kubernetes RBACing.
> 
> `clusterUser` kubeconfig: Cluster user kubeconfig is setup to authenticate and authorize via AAD. By default, an AAD user has no access to the cluster and authorization to an AKS cluster will only succeed if AKS cluster admin has setup the right role bindings for the user.
> 
> **In summary** yes, az aks get-credentials will return a kubeconfig (for `clusterUser` role) that `kubectl` will connect with credentials unique to each AAD user logging in via device login the first time a `kubectl` command is run or when AAD token expires. So `clusterUser` kubeconfig will not return shared credentials that are the same for all users. As long as the associated AAD group is updated to remove the user or if user leaves that organization they will not be able to access the cluster.
> 
> Also, as mentioned above `clusterAdmin` will have full access to the cluster like it does today (this can also be RBACed as described above). This is required to bootstrap the cluster for other AAD users for `clusterUser` roles.

So what is the AZURE RBAC on the AKS cluster (IAM) that will allow a user to use the `--admin` option or using a POST action to `https://management.azure.com/subscriptions/${SUB_ID}/resourcegroups/${RG_NAME}/providers/Microsoft.ContainerService/managedClusters/${AKS_NAME}/accessProfiles/${ROLE_NAME}/listcredential?api-version=2018-03-31`  and get credentials.

Can a user with only "reader" access to the AKS Cluster (Az management pane) use POST to `https://management.azure.com/subscriptions/${SUB_ID}/resourcegroups/${RG_NAME}/providers/Microsoft.ContainerService/managedClusters/${AKS_NAME}/accessProfiles/${ROLE_NAME}/listcredential?api-version=2018-03-31` and get credentials?

**Comment by alanjcastonguay** [comment]
https://docs.microsoft.com/en-us/azure/aks/certificate-rotation asserts the new `az aks rotate-certs` will rotate the cluster CA. This has the effect of the cluster no longer trusting certificates signed by the former CA, including that embedded in the `clusterAdmin` kubeconfig fetched with ` az aks get-credentials --admin`. The document suggests a 30 minute outage, so we probably don't want to rotate it frequently and still need to use Azure IAM rules to prevent most users from `listClusterAdminCredential`.

**Comment by ebr** [comment]
This behaviour is still very unclear to me.

I just performed the following test:

- Create a user in Azure AD, and grant it the `Contributor` role to my subscription
- `az login` as the user
- `az aks get-credentials`, specifying my AKS cluster
- observe `.kube/config` being updated with the cluster details, and the context being changed, as expected:
    ```
     - name: clusterUser_{myResourceGroup}_{myClusterName}
       user:
         client-certificate-data: {...}
         client-key-data: {...}
         token: {...}
   ```
- `kubectl get node` returns nodes, etc. I have access to the cluster
- **delete** the role binding from Azure AD. the Test User now has NO role.
- confirm this effect by `az account list --refresh` which returns a blank list now, and `az aks list` which returns that my access is denied
- **however!** `kubectl` still has access to the cluster, even 30 minutes after Azure AD access was revoked for the user. I am able to get nodes, pods, etc.

So this means the `clusterUser` credentials are still active. This seems like a massive security hole. Someone could walk out with their `.kube/config` and maintain access to the cluster... indefinitely???

Can anyone point me to documentation that specifically explains how to correctly integrate Azure AD with AKS RBAC so that disabling a user in AD also propagates to all AKS clusters they might have authenticated to?

**Comment by zuehlke-dea** [comment]
I think you get the credentials with --admin. 

This is how it should look like
First is and admin user. the second one is a normal user without client cert, but with AD login

`- name: clusterAdmin
  user:
    client-certificate-data: ...
    client-key-data: ...
    token: ...
- name: clusterUser
  user:
    auth-provider:
      config:
        access-token: ...
        apiserver-id: ...
        client-id: ...
        environment: AzurePublicCloud
        expires-in: ..
        expires-on: ..
        refresh-token: ...
        tenant-id: ...
      name: azure`

**Comment by ebr** [comment]
> are you sure you do not get the credentials with --admin?

@zuehlke-dea I am 100% positive about that.... i also made sure to remove all traces of previous authentication from my kube config.

I'm using `az` v2.0.80; apparently it's up to v2.1.0 as of last week, so I'm going to try that version.

**Comment by ebr** [comment]
- Upgraded to `az` CLI v2.1.0. 
- Cleaned out my kube config. 
- Created a separate user to test with. 
- ran `az login` as the new user
- ran `az aks get-credentials... ` certainly without the `--admin` flag

.... same results. I'm getting the `client-certificate` stanza in my config.

What am I doing wrong?

**Comment by flamingboo** [comment]
@ebr 
Have you ensure that your user has been assigned the role below?
 az role assignment create --assignee $user_objectid --role "Azure Kubernetes Service Cluster User Role" --scope $AKS_ID

**Comment by zuehlke-dea** [comment]
@ebr 
I just read your post again. And I am currently not sure that you do the correct things in general.

Do you follow the guide here https://docs.microsoft.com/en-us/azure/aks/azure-ad-integration ?

You need to:

1. Create a AKS with RBAC enabled
2. Create a server and a client SP following the guide above.
3. You do not login with az login. When you followed the guide above you login with your normal ad account, get the credentials with the comman (without --admin) and get the config I described. 
When you have a non RBAC enabled cluster you do not need the --admin to get those client certs (at least it was the case when we have done this last time)
4. When you got your credentials and use a kubectl command you get the hint to login and a device key which you have to enter in your browser
`To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code XXXXXXX to authenticate.`

then it should work.

**Comment by leyshon** [comment]
@ebr have you followed the steps here  https://docs.microsoft.com/en-us/azure/aks/azure-ad-integration-cli ?It sounds like you’re expecting that by removing the role that allows you to download the certificate from a user that this will stop them accessing the cluster. That’s simply not the case, all that role does is allow them to download the certificate. Once they have it they have it forever. You need to follow the steps in the link to use RBAC on the cluster. Once you have it enabled and type `kubectl get nodes`you’ll be prompted to log in to Azure AD. If that is happening then you can disable a user in Azure AD and they won’t be able to access the cluster any more.

**Comment by ebr** [comment]
Thank you @zuehlke-dea, @leyshon . RBAC was the missing piece. I thought I had it enabled, but I didn't! After deleting the cluster and re-creating one with RBAC, this works as you describe. Many thanks.

**Comment by zuehlke-dea** [comment]
yw

**Comment by vishal8k** [comment]
What about the question "How to disable --admin" ??? I just looked into the built-in cluster roles (cluster-admin) and cluster role bindings (aks-cluster-admin-binding) and tried editing/deleting them but still didn't work. The role bindings are automatically created again after deletion and the role remains deleted. but somehow I'm still having admin access to the cluster. What role is being actually used here? This cluster binding which is being created automatically after deletion is now pointing to a role that does not even exist but I'm still having admin access. What exactly is happening in the background with respect to Kubernetes  (not config files) when we use "--admin" flag.

**Comment by TomGeske** [comment]
--admin is controlled by Azure RBAC role [(azure-kubernetes-service-cluster-admin-role)](https://docs.microsoft.com/en-us/azure/role-based-access-control/built-in-roles#azure-kubernetes-service-cluster-admin-role). It basically ignores AAD and uses client certificates.
